File: dagger/pulumi/dagger.go
package pulumi

import (
	"context"
	_ "embed"
	"encoding/json"
	"fmt"

	"dagger.io/dagger"
)

//go:embed scripts/up.sh
var UpScript string

type Config struct {
	Version              string
	Runtime              string
	Stack                string
	StackCreate          bool
	AccessToken          string
	Passphrase           string
	EnvironmentVariables map[string]string
	ProgramDir           dagger.DirectoryID
	GCloudGke            bool
}

type Outputs map[string]interface{}

func Up(config Config, ctx context.Context, client *dagger.Client) (Outputs, error) {
	entrypointDir := client.Directory().WithNewFile("up.sh", dagger.DirectoryWithNewFileOpts{
		Contents: UpScript,
	})

	entrypointDirId, err := entrypointDir.ID(ctx)
	if err != nil {
		return nil, err
	}

	var stackCreate string
	if config.StackCreate {
		stackCreate = "true"
	} else {
		stackCreate = "false"
	}

	pulumi := client.
		Container().
		From(fmt.Sprintf("pulumi/pulumi-%s:%s", config.Runtime, config.Version)).
		WithMountedDirectory("/entrypoint", entrypointDirId).
		WithMountedDirectory("/work", config.ProgramDir).
		WithWorkdir("/work").
		WithEntrypoint([]string{"bash"}).
		WithEnvVariable("PULUMI_RUNTIME", config.Runtime).
		WithEnvVariable("PULUMI_STACK", config.Stack).
		WithEnvVariable("PULUMI_STACK_CREATE", stackCreate)

	for key, value := range config.EnvironmentVariables {
		pulumi = pulumi.WithEnvVariable(key, value)
	}

	if config.GCloudGke {
		pulumi = pulumi.WithEnvVariable("GCLOUD_GKE", "YES")
	}

	result := pulumi.Exec(dagger.ContainerExecOpts{
		Args: []string{"/entrypoint/up.sh"},
	})
	_, err = result.ExitCode(ctx)
	if err != nil {
		return nil, err
	}

	output, err := result.File("/output/json").Contents(ctx)
	if err != nil {
		return nil, err
	}

	outputs := make(Outputs)
	err = json.Unmarshal([]byte(output), &outputs)
	if err != nil {
		return nil, err
	}

	return outputs, nil
}

File: dagger/pulumi/dagger.ts
import Client from "@dagger.io/dagger";
import { getSourceDir } from "../utils/index.js";
import { Directory } from "@dagger.io/dagger/dist/api/client.gen";

interface Config {
  version: string;
  runtime: "nodejs" | "go";
  stack: string;
  programDirectory: Directory;

  stackCreate?: boolean;
  accessToken?: string;
  passphrase?: string;
  environmentVariables?: Record<string, string>;
  googleCloudGkeCluster?: boolean;
}

export const up = async (client: Client, config: Config): Promise<Object> => {
  const entrypointDir = client
    .host()
    .directory(`${getSourceDir(import.meta.url)}/scripts`);

  let pulumi = client
    .container()
    .from(`pulumi/pulumi-${config.runtime}:${config.version}`)
    .withMountedDirectory("/entrypoint", entrypointDir)
    .withMountedDirectory("/work", config.programDirectory)
    .withWorkdir("/work")
    .withEnvVariable("PULUMI_RUNTIME", config.runtime)
    .withEnvVariable("PULUMI_STACK", config.stack)
    .withEnvVariable(
      "PULUMI_STACK_CREATE",
      config.stackCreate ? "true" : "false"
    );

  for (const key in config.environmentVariables) {
    pulumi = pulumi.withEnvVariable(key, config.environmentVariables[key]);
  }

  if (config.googleCloudGkeCluster) {
    pulumi = pulumi.withEnvVariable("GCLOUD_GKE", "YES");
  }

  const result = await pulumi.withExec(["/entrypoint/up.sh"]);

  if ((await result.exitCode()) !== 0) {
    const stderr = await result.stderr();
    throw new Error(`Pulumi up failed: ${stderr}`);
  }

  const getOutputJson = await result.file("/output/json").contents();
  return JSON.parse(getOutputJson);
};

export const preview = async (
  client: Client,
  config: Config
): Promise<string> => {
  const entrypointDir = client
    .host()
    .directory(`${getSourceDir(import.meta.url)}/scripts`);

  let pulumi = client
    .container()
    .from(`pulumi/pulumi-${config.runtime}:${config.version}`)
    .withMountedDirectory("/entrypoint", entrypointDir)
    .withMountedDirectory("/work", config.programDirectory)
    .withWorkdir("/work")
    .withEnvVariable("PULUMI_RUNTIME", config.runtime)
    .withEnvVariable("PULUMI_STACK", config.stack)
    .withEnvVariable(
      "PULUMI_STACK_CREATE",
      config.stackCreate ? "true" : "false"
    );

  for (const key in config.environmentVariables) {
    pulumi = pulumi.withEnvVariable(key, config.environmentVariables[key]);
  }

  if (config.googleCloudGkeCluster) {
    pulumi = pulumi.withEnvVariable("GCLOUD_GKE", "YES");
  }

  const result = await pulumi.withExec(["/entrypoint/preview.sh"]);

  if ((await result.exitCode()) !== 0) {
    const stderr = await result.stderr();
    throw new Error(`Pulumi up failed: ${stderr}`);
  }

  return await result.stdout();
};
